### PR TITLE
166.m - implement character remapping

### PR DESCRIPTION
**Pull request by d-w-moore**

### CODE REVIEW

File: irods_capability_automated_ingest/sync_task.py [hunk 1]
```diff
@@ -284,6 +285,9 @@ def sync_entry(self, meta, cls, datafunc, metafunc):
             else:
                 target2 = syncer.construct_path(meta2, path)
 
+            char_map_func = getattr(event_handler.get_module(),'character_map',None)
```
Comment (trel) [inline irods_capability_automated_ingest/sync_task.py:288]: please add a comment about why this is the way it is...

aka... if the character map is defined, use it here, and set the new target.

File: irods_capability_automated_ingest/char_map_util.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import base64
+import hashlib
+
+#translate_string = str.translate
+from .process_map import translate_string
```
Comment (trel) [inline irods_capability_automated_ingest/char_map_util.py:5]: what have an import from an import?    can't we just add our one/two functions to utils.py?

**Review by trel — COMMENTED** [review]
excited about a demo!

File: irods_capability_automated_ingest/examples/use_character_map.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from irods_capability_automated_ingest.core import Core 
+from irods_capability_automated_ingest.utils import Operation
+import re
+
+class event_handler(Core):
+    re_non_alphanum = re.compile('[^a-zA-Z0-9]')      
+    @staticmethod                                     
+    def character_map():                              
+        return {event_handler.re_non_alphanum:'_'}    # map any non-ascii or non-alphanumeric
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods_capability_automated_ingest/examples/use_character_map.py:9]: the beauty of this approach is internally everything is cast to an OrderedDict.  That way you *can* pass an array of tuples to the constructor but (esp if you don't care about order) you don't have to.

File: irods_capability_automated_ingest/char_map_util.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        hash_suffix = ''
+        if new_el != el:
+            h = hashlib.sha1()
+            h.update(el.encode('utf8'))
+            hash_suffix = '-'+base64.b32encode(h.digest()).decode('utf8')
+        t_elem.append( new_el + hash_suffix )
+    return '/'.join(t_elem)
+
+if __name__=='__main__':
+    print(encode_path('/a/b'))
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods_capability_automated_ingest/char_map_util.py:23]: yes, will remove

File: irods_capability_automated_ingest/char_map_util.py (hunk 2 above)
Comment (d-w-moore) [reply irods_capability_automated_ingest/char_map_util.py:5]: agreed

**Comment by d-w-moore** [comment]
A keen observer might notice that we're being very inefficient by always encoding the original string using SHA1 to get the "unique-making suffix"... and they'd be right.  The suffix we encode could be potentially much shorter and store only the diff between original and transformed string.... e.g. we could use something like:
```
ascii_encode = base64.b32encode # a decent default if as many as 32 alpha-nums "map to themselves"
diff = [ x[0] for x in zip(orig_objname,transformed_objname) if x[0] != x[1] ]
suffix = ascii_encode(b"".join( [u.encode('utf8') for u in diff] ))

```

File: irods_capability_automated_ingest/process_map.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import re
+import collections
+
+_regex = re.compile('')
+string_replace = lambda s,m: s.translate({ord(k):v for k,v in m.items()})
```
Comment (d-w-moore) [inline irods_capability_automated_ingest/process_map.py:5]: hm... probably want an `if v is not None` in there.  Allowing the deletion of characters won't have a good effect on uniqueness.

File: irods_capability_automated_ingest/examples/use_character_map.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (trel) [reply irods_capability_automated_ingest/examples/use_character_map.py:9]: Gotcha - yes OrderedDict seems sufficient.

This ingest tool is Python3.7+ only, so we can rely on known good behavior of OrderedDict...

**Comment by d-w-moore** [comment]
We are considering whether to remove (or deprecate?) the feature introduced in this commit: ed8d794 . (
The original issue solved by that commit is here: https://github.com/irods/irods_capability_automated_ingest/issues/40#issue-354053528 .)

It may make sense to do that now, since we are not certain how the character remapping (the subject of the discussion within this pull request) should interact with that feature....

In that issue's problem description, it is mentioned that certain characters cause a `UnicodeEncodeError` when the directory scanner (`os.scandir`) turns up a filename or path element outside the valid UTF-8 coding space. An example would be a file created with, for example, `open(b"name"+bytes([B]),'w')` where 128 <= B < 256.
In modern OS'es and Python versions, `os.scandir` seems to use the 'surrogateescape' error handler when translating such filenames, so that a filename with binary encoding `b'name\x80'` becomes `u'name\udc80'` when decoded using UTF-8 (and thus iRODS will register it without incident). This theory is corroborated by the fact that the `UnicodeEncodeError` is now seemingly impossible to reproduce on any OS or Python version supported by the automated ingest.
It's for this reason I would suggest removal of the special handling of "illegal UTF-8" introduced in ed8d794 ....Thoughts @alanking @korydraughn  @trel @SwooshyCueb

**Comment by d-w-moore** [comment]
To lay a little more solid foundation for the post above, run the script here and observe the output.  All those weird final characters (byte 128-255) are translated to their Unicode surrogate codepoints \uDC80-\uDCFF , and that is shown in the output of the script.
```python
#!/usr/bin/python3
import os
import tempfile

xdir = tempfile.mkdtemp(dir=b'.')

unicode_xdir = xdir.decode('utf8')

for b in (bytes([x]) for x in range(126,256)):
    with tempfile.NamedTemporaryFile(dir=xdir,suffix=b'_'+b):
        _file = next(os.scandir(unicode_xdir))
        print(repr( _file.name ))

```

**Comment by d-w-moore** [comment]
What I'm assuming, and what I suppose is the thing needs to be thoroughly tested, is that all modern filesystems and their subobjects like to be queried as Unicode (UTF-8 encoded) entities, and that when we circumvent that by directly constructing filenames as bytestrings, the appropriate characters are given the surrogate value when they don't fit with the UTF-8 encoding.

File: irods_capability_automated_ingest/sync_task.py [hunk 6]
```diff
@@ -298,9 +298,12 @@ def sync_entry(self, meta, cls, datafunc, metafunc):
 
             if 'unicode_error_filename' not in meta:
                 if char_map_func:
-                    target2 = translate_path(target2,char_map_func())
-                    meta2["b64_reason"] = "character_map"
-                    meta2['b64_path_str_X'] = b64_path_str
+                    temp = translate_path(target2,char_map_func())
+                    # arrange for AVU to be attached only when logical name changes
+                    if temp != target2:
```
Comment (trel) [inline irods_capability_automated_ingest/sync_task.py:303]: maybe 'translated_path' rather than temp...

Comment (d-w-moore) [reply irods_capability_automated_ingest/sync_task.py:303]: right

File: README.md [hunk 7]
```diff
@@ -100,6 +101,45 @@ These options are used at the "start" of an ingest job.
...
+            }
+        # ...
+```
+The returned dictionary, in this case, indicates that the ingest process should replace all non-alphanumeric (as
+well as non-ASCII) characters with an underscore wherever they may occur in an otherwise normally generated logical path.
+The substition process also applies to the intermediate (ie collection name) elements in a logical path, and a suffix is
+appended to affected path elements to avoid potential collisions with other remapped object names.
+
+Each key of the returned dictionary may be a (1) character, (2) tuple of characters, (3) regular expression, or (3) a callable
+which accepts a character argument and returns a boolean indicating whether that character is one that should be remapped.
```
Comment (trel) [inline README.md:123]: might show an example of each of these four...

Comment (trel) [reply README.md:123]: and number four should not be called number three

File: README.md [hunk 8]
```diff
@@ -100,6 +101,45 @@ These options are used at the "start" of an ingest job.
...
+
+In the event that the order-of-substitution is significant, the method may instead return a list of key-value tuples.
+
+### UnicodeEncodeError
+Any file whose path in the filesystem whose ingest results in a UnicodeEncodeError exception being raised (e.g. by the
```
Comment (trel) [inline README.md:128]: let's mention that this happens regardless of event handler code, and that it will be checked/executed prior to any character_map that may be defined.

File: irods_capability_automated_ingest/test/test_irods_sync.py [hunk 9]
```diff
@@ -1492,21 +1522,45 @@ def run_scan_with_event_handler(self, eh_name, job_name = DEFAULT_JOB_NAME):
...
+            if self.DETAILED_CHECK:
+                self.assert_physical_path_and_resource(session, expected_physical_path)
+                self.assert_data_object_contents(session)
+                self.assert_data_object_size(session)
+
+#---------------------------------------------
+
+class _supplementary_tests_with_bad_filename:
+
+    @unittest.skip("SYS_RESC_DOES_NOT_EXIST - index out of range on replica - DWM 09/12/2022")
```
Comment (trel) [inline irods_capability_automated_ingest/test/test_irods_sync.py:1563]: no DWM / dates in the skipinfo

File: irods_capability_automated_ingest/test/test_irods_sync.py [hunk 10]
```diff
@@ -1605,6 +1644,49 @@ def test_put_append(self):
...
+# This is the transform built into the example files:
+# (Exclusion of '/' added here to handle path-at-a-time)
+
+regex = re.compile('[^0-9a-zA-Z/]')
+
+def character_mapping_transform(this):
+        this.remapped_bad_filename = regex.sub('_',this.BAD_FILENAME.decode('utf8'))
+        this.dest_coll_path = regex.sub('_',this.dest_coll_path)
+
+# TODO base64 transform should be the same for each of the current types of test - DWM
```
Comment (trel) [inline irods_capability_automated_ingest/test/test_irods_sync.py:1659]: no DWMs

File: irods_capability_automated_ingest/test/test_irods_sync.py [hunk 11]
```diff
@@ -1605,6 +1644,49 @@ def test_put_append(self):
...
+
+def b64_calculation(this):
+    utf8_escaped_abspath = this.bad_filepath
+    this.b64_path_str = base64.b64encode(utf8_escaped_abspath)
+
+
+class Test_irods_sync_character_mapped_path(_Test_irods_sync_with_bad_filename,
+                                            unittest.TestCase):
+    # These tests will operate on a "bad filename" wherein characters are in need of re-mapping
+    BAD_FILENAME = b'test-file~with@5!non.alphas'  # maybe add some unicode - DWM
```
Comment (trel) [inline irods_capability_automated_ingest/test/test_irods_sync.py:1669]: no DWMs

File: irods_capability_automated_ingest/test/test_irods_sync.py [hunk 12]
```diff
@@ -1605,6 +1644,49 @@ def test_put_append(self):
...
+
+
+# This is the transform built into the example files:
+# (Exclusion of '/' added here to handle path-at-a-time)
+
+regex = re.compile('[^0-9a-zA-Z/]')
+
+def character_mapping_transform(this):
+        this.remapped_bad_filename = regex.sub('_',this.BAD_FILENAME.decode('utf8'))
+        this.dest_coll_path = regex.sub('_',this.dest_coll_path)
```
Comment (trel) [inline irods_capability_automated_ingest/test/test_irods_sync.py:1657]: indentation

File: irods_capability_automated_ingest/test/test_irods_sync.py [hunk 13]
```diff
@@ -1605,6 +1644,49 @@ def test_put_append(self):
...
+    EH_SUFFIX = '_using_char_map'
+    B64_CALCULATION = b64_calculation
+    CHARACTER_MAPPING_TRANSFORM = character_mapping_transform
+    LOGICAL_PATH_CALCULATION = \
+        lambda this : join(this.dest_coll_path, this.remapped_bad_filename)
+    ALLOW_LOGICAL_NAME_SUFFIX = True
+    DETAILED_CHECK = False
+    ANNOTATION_REASON = 'character_map'
+
+#-----------------------------------------------------------
```
Comment (trel) [inline irods_capability_automated_ingest/test/test_irods_sync.py:1679]: remove

File: docker/ingest-test/docker-compose.yml [hunk 14]
```diff
@@ -23,6 +23,14 @@ services:
...
     build:
       context: provider
+      args:
+              irods_version: 4.3.0-1~bionic
+              irods_version_major_minor: 4.3
+              py_version: 3
+              # For older iRODS server:
+              #irods_version: 4.2.11-1~bionic
+              #irods_version_major_minor: 4.2
+              #py_version: ""
```
Comment (trel) [inline docker/ingest-test/docker-compose.yml:33]: mergeable?   or remove

File: docker/ingest-test/icommands.env [hunk 15]
```diff
@@ -4,5 +4,7 @@ IRODS_USER_NAME=rods
 IRODS_ZONE_NAME=tempZone
 IRODS_ENVIRONMENT_FILE=/irods_environment.json
 IRODS_PASSWORD=rods
-PIP_PACKAGE=git+https://github.com/irods/irods_capability_automated_ingest@main
+#PIP_PACKAGE=git+https://github.com/irods/irods_capability_automated_ingest@main
```
Comment (trel) [inline docker/ingest-test/icommands.env:7]: remove dead code

File: docker/ingest-test/icommands.env [hunk 16]
```diff
@@ -4,5 +4,7 @@ IRODS_USER_NAME=rods
...
+PIP_PACKAGE=git+https://github.com/d-w-moore/irods_capability_automated_ingest@166.m
 TEST_CASE=irods_capability_automated_ingest.test.test_irods_sync
+#TEST_CASE=irods_capability_automated_ingest.test.test_irods_sync.Test_irods_sync_UnicodeEncodeError.test_register
```
Comment (trel) [inline docker/ingest-test/icommands.env:10]: remove dead code

File: irods_capability_automated_ingest/test/test_irods_sync.py [hunk 17]
```diff
@@ -1492,21 +1522,45 @@ def run_scan_with_event_handler(self, eh_name, job_name = DEFAULT_JOB_NAME):
...
+
+        with iRODSSession(**get_kwargs()) as session:
+            self.assert_logical_path(session, allow_suffix = self.ALLOW_LOGICAL_NAME_SUFFIX )
+            self.assert_metadata_annotation(session)
+            if self.DETAILED_CHECK:
+                self.assert_physical_path_and_resource(session, expected_physical_path)
+                self.assert_data_object_contents(session)
+                self.assert_data_object_size(session)
+
+#---------------------------------------------
```
Comment (trel) [inline irods_capability_automated_ingest/test/test_irods_sync.py:1559]: remove comment

File: docker/ingest-test/docker-compose.yml [hunk 18]
```diff
@@ -23,6 +23,14 @@ services:
   irods-catalog-provider:
     build:
       context: provider
+      args:
+              irods_version: 4.3.0-1~bionic
+              irods_version_major_minor: 4.3
+              py_version: 3
```
Comment (trel) [inline docker/ingest-test/docker-compose.yml:29]: indentation

File: README.md [hunk 19]
```diff
@@ -100,6 +101,45 @@ These options are used at the "start" of an ingest job.
...
+                re.compile('[^a-zA-Z0-9]'):'_'
+            }
+        # ...
+```
+The returned dictionary, in this case, indicates that the ingest process should replace all non-alphanumeric (as
+well as non-ASCII) characters with an underscore wherever they may occur in an otherwise normally generated logical path.
+The substition process also applies to the intermediate (ie collection name) elements in a logical path, and a suffix is
+appended to affected path elements to avoid potential collisions with other remapped object names.
+
+Each key of the returned dictionary may be a (1) character, (2) tuple of characters, (3) regular expression, or (3) a callable
```
Comment (trel) [inline README.md:122]: please include an example for all four.

and make sure the second number three is number four.

File: README.md [hunk 20]
```diff
@@ -100,6 +101,62 @@ These options are used at the "start" of an ingest job.
...
+       ('\\','#','-'): '\u2093'  # any character of the tuple is replaced by a Unicode small script x
+   ```
+   3. regular expression
+   ```
+       re.compile('[\u0100-\U0010ffff]'): '_'  # any character outside of range 0-256 becomes an underscore
+   ```
+   4. callable accepting a character argument and returning a boolean
+   ```
+       (lambda c: ord(c) in range(ord('z')+1,128)): ':'  # ASCII codes above 'z' become ':'
+   ```
```
Comment (trel) [inline README.md:140]: i like it - please put the comment above each example

e.g.
```
# any character of the tuple is replaced by a Unicode small script x
('\\','#','-'): '\u2093'  
```

File: README.md [hunk 21]
```diff
@@ -100,6 +101,62 @@ These options are used at the "start" of an ingest job.
...
+well as non-ASCII) characters with an underscore wherever they may occur in an otherwise normally generated logical path.
+The substition process also applies to the intermediate (ie collection name) elements in a logical path, and a suffix is
+appended to affected path elements to avoid potential collisions with other remapped object names.
+
+Each key of the returned dictionary indicates a character or set of characters needing substitution.
+Possible key types include:
+
+   1. character
+   ```
+       '\\':'_'	 # substitute backslashes with underscores
```
Comment (trel) [inline README.md:127]: space after colon?

File: docker/ingest-test/docker-compose.yml (hunk 14 above)
Comment (d-w-moore) [reply docker/ingest-test/docker-compose.yml:33]: Mergeable and maybe important to show that tests may be run on 4.2 as well as 4.3 . Does this require more comments to make things more explicit?

Comment (korydraughn) [reply docker/ingest-test/docker-compose.yml:33]: This can be documented in a README that lives inside of docker/ingest-test.

IMO, yes this needs more words. Right now, people can only assume what these do. We need to do our best to take the guess work out of things. That doesn't mean we have to document every single thing, but providing words as to what each parameter is for and then providing an example of usage is probably good enough.

**Comment by trel** [comment]
one indentation comment left

and then resolve the ones that are done.   looking pretty good.

File: irods_capability_automated_ingest/test/test_irods_sync.py (hunk 12 above)
Comment (d-w-moore) [reply irods_capability_automated_ingest/test/test_irods_sync.py]: done.

**Comment by d-w-moore** [comment]
Putting it through one more test before merge.

**Comment by trel** [comment]
Click resolve above on two comments and squash when ready.   No # yet.

**Comment by d-w-moore** [comment]
Squashed with no #

File: irods_capability_automated_ingest/sync_task.py [hunk 22]
```diff
@@ -284,7 +291,23 @@ def sync_entry(self, meta, cls, datafunc, metafunc):
             else:
                 target2 = syncer.construct_path(meta2, path)
 
+            # If the event handler has a character_map function, it should have returned a
+            # structure (either a dict or a list/tuple of key-value tuples) to be used for
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods_capability_automated_ingest/sync_task.py:295]: It won't be a silent failure , but it won't be pretty either. Should  an explicit error could be thrown, which might make things friendlier?

[the remark this answers is not in the document]
Comment (trel) [reply irods_capability_automated_ingest/sync_task.py:295]: Yes, we should check for a sane/valid value and STOP any work being done if it's not.

The error should say that the return type of the character_map was invalid.

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods_capability_automated_ingest/sync_task.py:295]: Sounds like we should check that up front then, at the moment the --event_handler option is submitted to the `irods_sync.py` script. Sound like the right approach, @alanking ?

[the remark this answers is not in the document]
Comment (korydraughn) [reply irods_capability_automated_ingest/sync_task.py:295]: If we can detect issues on startup, then the rest of the code becomes faster because it can assume things are in a good state. That also means the rest of the code must be implemented so that invalid values do not appear as a result of carrying out various operations.

So overall, yes.

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods_capability_automated_ingest/sync_task.py:295]: I've made an attempt, want to look it over?  Have tested manually, too.  Does it need its own unit test ?

File: irods_capability_automated_ingest/irods_sync.py [hunk 23]
```diff
@@ -76,6 +101,8 @@ def handle_start(args):
     if ex_file_arg != None:
         ex_arg_list = [x.strip() for x in ex_file_arg[0].split(',')]
 
+    check_event_handler(args.event_handler)
```
Comment (trel) [inline irods_capability_automated_ingest/irods_sync.py:104]: this will run once per sync job, not once per file-to-be-synced, yes?

File: irods_capability_automated_ingest/irods_sync.py [hunk 24]
```diff
@@ -43,6 +43,31 @@ def get_celery_broker_info():
     return host, port, db
 
 
+class character_map_argument_error(Exception): pass
+
+def check_event_handler(filename):
```
Comment (trel) [inline irods_capability_automated_ingest/irods_sync.py:48]: please put the comment from the commit message here in the code...
- the goal, the reason, the what.

File: irods_capability_automated_ingest/sync_task.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (trel) [reply irods_capability_automated_ingest/sync_task.py:295]: left comments

File: irods_capability_automated_ingest/irods_sync.py (hunk 23 above)
Comment (d-w-moore) [reply irods_capability_automated_ingest/irods_sync.py:104]: Right. Actually at the command line invocation level, before anything starts, an exception will be thrown if the named event handler is not  correct.

File: irods_capability_automated_ingest/sync_task.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods_capability_automated_ingest/sync_task.py:295]: Good.  And yes, tests passing as designed on 4.2.11 and 4.3.0

File: irods_capability_automated_ingest/irods_sync.py (hunk 23 above)
Comment (trel) [reply irods_capability_automated_ingest/irods_sync.py:104]: perfect.

**Comment by trel** [comment]
and squash again when ready.  still no #.

**Comment by d-w-moore** [comment]
Ok, done.

File: irods_capability_automated_ingest/irods_sync.py (hunk 23 above)
Comment (d-w-moore) [reply irods_capability_automated_ingest/irods_sync.py:110]: yes

File: .gitignore [hunk 25]
```diff
@@ -0,0 +1,3 @@
+*.pyc
+*egg-info/
+cscope.*
```
Comment (trel) [inline .gitignore:3]: maybe create a housekeeping issue for docs/typos/configs

Comment (d-w-moore) [reply .gitignore:3]: ok, done.

**Review by trel — APPROVED** [review]
looks good to me.  let's await @alanking's eyes one last time before #.

**Review by alanking — APPROVED** [review]

**Comment by trel** [comment]
alright dan, please # em.

**Comment by d-w-moore** [comment]
done!
